src/display.rs

```diff
@@ -420,3 +481,87 @@ impl Display {
         self.window().send_xim_spot(nspot_x, nspot_y);
     }
 }
+
+// Check if the underline/strikeout state has changed
+// This returns a new rectangle whenever an underline/strikeout ends
+fn calculate_cell_line_state(
+    cell: RenderableCell,
+    start_cell_line: &mut Option<RenderableCell>,
+    last_cell: &Option<RenderableCell>,
+    metrics: &Metrics,
+    size_info: &SizeInfo,
+    flag: cell::Flags,
+) -> Option<(Rect<u32>, Rgb)> {
+    match *start_cell_line {
+        // If line is already started, check for end
+        Some(start) => {
+            // No change in line
+            if cell.line == start.line
+                && cell.flags.contains(flag)
+                && cell.fg == start.fg
+            {
+                return None;
+            }
+
+            // Check if we need to start a new line
+            *start_cell_line = if cell.flags.contains(flag) {
+                // Start a new line
+                Some(cell)
+            } else {
+                // Disable line
+                None
+            };
+
+            return Some(
+                cell_line_rect(&start, &last_cell.unwrap(), metrics, size_info, flag)
+            );
+        },
+        // Check for new start of line
+        None => if cell.flags.contains(flag) {
+            *start_cell_line = Some(cell);
+        },
+    }
+    None
+}
+
+// Create a colored rectangle for an underline/strikeout based on two cells
+fn cell_line_rect(
+    start: &RenderableCell,
+    end: &RenderableCell,
+    metrics: &Metrics,
+    size: &SizeInfo,
+    flag: cell::Flags,
+) -> (Rect<u32>, Rgb) {
+    let x = (start.column.0 as f64 * metrics.average_advance) as u32;
+    let end_x = ((end.column.0 + 1) as f64 * metrics.average_advance) as u32;
+    let width = end_x - x;
+
+    let (y, height) = match flag {
+        cell::Flags::UNDERLINE => {
+            // Get the baseline positon and offset it down by (-) underline position
+            // then move it up by half the underline thickness
+            let y = ((start.line.0 as f32 + 1.) * metrics.line_height as f32 + metrics.descent
+                - metrics.underline_position
+                - metrics.underline_thickness / 2.)
+                .round() as u32;
```

Comment:
Just out of curiosity I replaced `.round()` with `.floor()` here and got these results:
- pixel perfect underline on font sizes [8-18, 21]
- underline beyond cell height on font sizes [5-7, 19-20]

This is obviously not a solution, but it got me wondering: `metrics.*` properties are all rounded in `font/src/ft/mod.rs`, here we have some calculations based on already rounded values and then round again, can't this introduce some increasing calculation error? What if `metrics.*` were not rounded in `font/src/ft/mod.rs`, but left as `f32`, and here we had rounding only once, in the end of all calculations?

Reply:
In case there are multiple Consolas fonts with different metrics, I use [this one](https://aur.archlinux.org/packages/ttf-consolas-with-powerline/) - try, maybe you can reproduce with it 🙂

Reply:
I think rounding multiple times should be fine because these metrics should never be float values. The underline position for example shouldn't be `11.2` pixels below the baseline because the concept of `0.2` pixels doesn't exist. And after we make use of the metrics we have to round again because we need `u32`s, so it's either floor, ceil or round to get valid integers.

I think that approach is correct, however if someone else knows more, please correct me.

Reproducing your issue will probably be hard for me because of potential DPI differences. I've used the `consolas-fonts` package to test though, so they might be different.

The question is just if my "algorithm" is correct really. I can't see anything wrong with it and it should be fine with the values you've supplied. If the algorithm is correct, but the font still specifies an offset that is out of bounds, I think alacritty should handle it.

I'm gonna go ahead and implement a bounds check which will make sure that whenever the underline is out of bounds, the underline will be moved up so it's exactly within bounds (using the thickness provided with the font). This should fix the issue and also help with potentially broken font metrics.

However if my approach to this calculation is wrong and maybe rounding shouldn't take place in the metrics, we should still fix this instead of just relying on a bounds check to catch off-by-one errors.

Reply:
I added this to your patch:

```
println!("CALC POS F32: {}", f32::from(face.ft_face.underline_position()) * x_scale / 64.);
println!("CALC THI F32: {}", f32::from(face.ft_face.underline_thickness()) * x_scale / 64.);
```

Let's look at a random output:

```
-----------
CELL_HEIGHT: 44
DESCENT: -10
UNDERLINE POS: -482
UNDERLINE THI: 144
X SCALE: 1.1733398
CALC POS F32: -8.836716
CALC THI F32: 2.6400146
CALC POS: -9
CALC THI: 3
-----------
```

What I was saying is that because F32 values can have any fractional part, the following hypothetical example can happen:

```
CALC_POS_F32 = 7.49
CALC_THI_F32 = 2.49

(CALC_POS_F32.round() + CALC_THI_F32.round() / 2.0).round() == (7 + 2 / 2).round() == 8
(CALC_POS_F32 + CALC_THI_F32 / 2.0).round() == (8.735).round() == 9
```

Let me know if this makes sense, I'll try the new bounds check in meantime.

Reply:
Yeah I understand what you mean. But in my opinion it doesn't make any sense to have a fractional underline position/thickness. It should always be an integer offset.

Reply:
I don't see any problem with your algorithm either, and just out of curiosity I verified that simply removing `.round()` when calculating `CALC POS` and `CALC THI` doesn't solve my particular problem, so I agree, let's keep the code as it is. Bounds check on the other hand is solving the problem well 🙂